ir/instr.cpp

```diff
@@ -1911,4 +1911,125 @@ unique_ptr<Instr> Memcpy::dup(const string &suffix) const {
   return make_unique<Memcpy>(*dst, *src, *bytes, align_dst, align_src, move);
 }
 
+
+vector<Value*> ExtractElement::operands() const {
+  return { v, idx };
+}
+
+void ExtractElement::rauw(const Value &what, Value &with) {
+  RAUW(v);
+  RAUW(idx);
+}
+
+void ExtractElement::print(ostream &os) const {
+  os << getName() << " = extractelement " << *v << ", " << *idx;
+}
+
+StateValue ExtractElement::toSMT(State &s) const {
+  auto &[iv, ip] = s[*idx];
+  auto vty = v->getType().getAsAggregateType();
+  expr inbounds = !iv.uge(vty->numElementsConst());
+  auto [rv, rp] = vty->extract(s[*v], iv);
+  return { move(rv), ip && inbounds && rp };
+}
+
+expr ExtractElement::getTypeConstraints(const Function &f) const {
+  return Value::getTypeConstraints() &&
+         v->getType().enforceVectorType([&](auto &ty)
+                                        { return ty == getType(); }) &&
+         idx->getType().enforceIntType();
+}
+
+unique_ptr<Instr> ExtractElement::dup(const string &suffix) const {
+  return make_unique<ExtractElement>(getType(), getName() + suffix, *v, *idx);
+}
+
+
+vector<Value*> InsertElement::operands() const {
+  return { v, e, idx };
+}
+
+void InsertElement::rauw(const Value &what, Value &with) {
+  RAUW(v);
+  RAUW(e);
+  RAUW(idx);
+}
+
+void InsertElement::print(ostream &os) const {
+  os << getName() << " = insertelement" << *v << ", " << *e << ", " << *idx;
+}
+
+StateValue InsertElement::toSMT(State &s) const {
+  auto &[iv, ip] = s[*idx];
+  auto vty = v->getType().getAsAggregateType();
+  expr inbounds = !iv.uge(vty->numElementsConst());
+  auto [rv, rp] = vty->update(s[*v], s[*e], iv);
+  return { move(rv), expr::mkIf(ip && inbounds,
+                                move(rp), expr::mkInt(-1, vty->bits())) };
+}
+
+expr InsertElement::getTypeConstraints(const Function &f) const {
+  return Value::getTypeConstraints() &&
+         getType() == v->getType() &&
+         v->getType().enforceVectorType([&](auto &ty)
+                                        { return ty == e->getType(); }) &&
+         idx->getType().enforceIntType();
+}
+
+unique_ptr<Instr> InsertElement::dup(const string &suffix) const {
+  return make_unique<InsertElement>(getType(), getName() + suffix,
+                                    *v, *e, *idx);
+}
+
+
+vector<Value*> ShuffleVector::operands() const {
+  return { v1, v2, mask };
+}
+
+void ShuffleVector::rauw(const Value &what, Value &with) {
+  RAUW(v1);
+  RAUW(v2);
+  RAUW(mask);
+}
+
+void ShuffleVector::print(ostream &os) const {
+  os << getName() << " = shufflevector " << *v1 << ", " << *v2 << ", " << *mask;
+}
+
+StateValue ShuffleVector::toSMT(State &s) const {
+  auto vty = v1->getType().getAsAggregateType();
+  auto mty = mask->getType().getAsAggregateType();
+  auto ty = getType().getAsAggregateType();
```

Comment:
you can inline this in the return statement.

Reply:
fixed :)